### ISSUE TITLE
Compiler bug in macos (apple-darwin), compiles and gives error message in playground

### ISSUE THREAD

**Issue by harrylaou** [rust-lang/rust#110803]
### Code

```Rust
use std::collections::HashMap;

use std::fs;
fn main() {
    // if env::args().len() < 1 {
    //     println!("Program requires as least 1 arguments.");
    //     return;
    // }

    // let filename = env::args().nth(1).unwrap();
    let filename = "earth_to_the_moon.txt";

    let text = fs::read_to_string(filename).unwrap();
    let contents = text.split_whitespace().map(|x| x.to_lowercase());

    let mut hash_map = HashMap::new();

    for word in contents {
        let count = hash_map.entry(word).or_insert(0);
        *count += 1;
    }

    fn compare(a: &(&str, &i32), b: &(&str, &i32)) -> std::cmp::Ordering {
        let (_, a_count) = a;
        let (_, b_count) = b;
        b_count.cmp(a_count)
    }

    let mut sorted = hash_map
        .iter()
        .collect::<Vec<_>>()
        .sort_by(compare)
        .take(10);

    // println!("The top 10 most flown missions are: {}", sorted);
}
```

### Meta

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021 gives a proper error message

`rustc --version --verbose`:
```
rustc 1.68.2 (9eb3afe9e 2023-03-27)
binary: rustc
commit-hash: 9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0
commit-date: 2023-03-27
host: x86_64-apple-darwin
release: 1.68.2
LLVM version: 15.0.6
```

### Error output

```
Compiling count_words v0.1.0 (/Users/harrylaou/myworkspace/rust/linkedin-rust-essential-training/count_words)
error: internal compiler error: compiler/rustc_infer/src/infer/region_constraints/mod.rs:568:17: cannot relate bound region: ReLateBound(DebruijnIndex(0), BoundRegion { var: 1, kind: BrNamed(DefId(0:9 ~ count_words[0ce7]::main::compare::'_#1), '_) }) <= '_#20r

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/compiler/rustc_errors/src/lib.rs:987:33
stack backtrace:
   0:        0x10cf816e6 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h47bed05ceb7970ee
   1:        0x10cfdefca - core::fmt::write::hefd823d99333384a
   2:        0x10cf73eac - std::io::Write::write_fmt::hb978956bb1b85e2b
   3:        0x10cf814ca - std::sys_common::backtrace::print::hc9e298c4664da113
   4:        0x10cf848c3 - std::panicking::default_hook::{{closure}}::hb97411d6e916e1d2
   5:        0x10cf84618 - std::panicking::default_hook::hc3ac81176817192f
   6:        0x116ba1d8a - rustc_driver[38920cb2344860e2]::DEFAULT_HOOK::{closure#0}::{closure#0}
   7:        0x10cf850b9 - std::panicking::rust_panic_with_hook::hc11209fa2686218a
   8:        0x11722faf7 - std[963c226d0575182f]::panicking::begin_panic::<rustc_errors[112d18aeec5719f8]::ExplicitBug>::{closure#0}
   9:        0x117225619 - std[963c226d0575182f]::sys_common::backtrace::__rust_end_short_backtrace::<std[963c226d0575182f]::panicking::begin_panic<rustc_errors[112d18aeec5719f8]::ExplicitBug>::{closure#0}, !>
  10:        0x11b030209 - std[963c226d0575182f]::panicking::begin_panic::<rustc_errors[112d18aeec5719f8]::ExplicitBug>
  11:        0x11730aa29 - std[963c226d0575182f]::panic::panic_any::<rustc_errors[112d18aeec5719f8]::ExplicitBug>
  12:        0x117307f48 - <rustc_errors[112d18aeec5719f8]::HandlerInner>::span_bug::<rustc_span[abe350e59e28df37]::span_encoding::Span, &alloc[622432b1fc1d0ab9]::string::String>
  13:        0x117307d5e - <rustc_errors[112d18aeec5719f8]::Handler>::span_bug::<rustc_span[abe350e59e28df37]::span_encoding::Span, &alloc[622432b1fc1d0ab9]::string::String>
  14:        0x1172890eb - rustc_middle[7044477369b49eec]::util::bug::opt_span_bug_fmt::<rustc_span[abe350e59e28df37]::span_encoding::Span>::{closure#0}
  15:        0x117289147 - rustc_middle[7044477369b49eec]::ty::context::tls::with_opt::<rustc_middle[7044477369b49eec]::util::bug::opt_span_bug_fmt<rustc_span[abe350e59e28df37]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  16:        0x1172884ef - rustc_middle[7044477369b49eec]::ty::context::tls::with_context_opt::<rustc_middle[7044477369b49eec]::ty::context::tls::with_opt<rustc_middle[7044477369b49eec]::util::bug::opt_span_bug_fmt<rustc_span[abe350e59e28df37]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  17:        0x117288491 - rustc_middle[7044477369b49eec]::util::bug::opt_span_bug_fmt::<rustc_span[abe350e59e28df37]::span_encoding::Span>
  18:        0x11b0301ff - rustc_middle[7044477369b49eec]::util::bug::span_bug_fmt::<rustc_span[abe350e59e28df37]::span_encoding::Span>
  19:        0x117281fa0 - <rustc_infer[6cab06927fa701c3]::infer::region_constraints::RegionConstraintCollector>::make_subregion
  20:        0x11727af82 - <rustc_infer[6cab06927fa701c3]::infer::region_constraints::RegionConstraintCollector>::make_eqregion
  21:        0x11728efda - <rustc_infer[6cab06927fa701c3]::infer::equate::Equate as rustc_middle[7044477369b49eec]::ty::relate::TypeRelation>::regions
  22:        0x117297470 - rustc_middle[7044477369b49eec]::ty::relate::super_relate_tys::<rustc_infer[6cab06927fa701c3]::infer::equate::Equate>
  23:        0x1172146e0 - <rustc_infer[6cab06927fa701c3]::infer::InferCtxt>::super_combine_tys::<rustc_infer[6cab06927fa701c3]::infer::equate::Equate>
  24:        0x117293289 - <rustc_infer[6cab06927fa701c3]::infer::equate::Equate as rustc_middle[7044477369b49eec]::ty::relate::TypeRelation>::tys
  25:        0x117260bc9 - <core[fd5e363bb4760a86]::result::Result<rustc_middle[7044477369b49eec]::ty::Ty, rustc_middle[7044477369b49eec]::ty::error::TypeError> as rustc_type_ir[12a6d7d457ec9c1a]::InternIteratorElement<rustc_middle[7044477369b49eec]::ty::Ty, rustc_middle[7044477369b49eec]::ty::Ty>>::intern_with::<core[fd5e363bb4760a86]::iter::adapters::map::Map<core[fd5e363bb4760a86]::iter::adapters::zip::Zip<core[fd5e363bb4760a86]::iter::adapters::copied::Copied<core[fd5e363bb4760a86]::slice::iter::Iter<rustc_middle[7044477369b49eec]::ty::Ty>>, core[fd5e363bb4760a86]::iter::adapters::copied::Copied<core[fd5e363bb4760a86]::slice::iter::Iter<rustc_middle[7044477369b49eec]::ty::Ty>>>, rustc_middle[7044477369b49eec]::ty::relate::super_relate_tys<rustc_infer[6cab06927fa701c3]::infer::equate::Equate>::{closure#2}>, <rustc_middle[7044477369b49eec]::ty::context::TyCtxt>::mk_tup<core[fd5e363bb4760a86]::iter::adapters::map::Map<core[fd5e363bb4760a86]::iter::adapters::zip::Zip<core[fd5e363bb4760a86]::iter::adapters::copied::Copied<core[fd5e363bb4760a86]::slice::iter::Iter<rustc_middle[7044477369b49eec]::ty::Ty>>, core[fd5e363bb4760a86]::iter::adapters::copied::Copied<core[fd5e363bb4760a86]::slice::iter::Iter<rustc_middle[7044477369b49eec]::ty::Ty>>>, rustc_middle[7044477369b49eec]::ty::relate::super_relate_tys<rustc_infer[6cab06927fa701c3]::infer::equate::Equate>::{closure#2}>>::{closure#0}>
  26:        0x117297509 - rustc_middle[7044477369b49eec]::ty::relate::super_relate_tys::<rustc_infer[6cab06927fa701c3]::infer::equate::Equate>
  27:        0x1172146e0 - <rustc_infer[6cab06927fa701c3]::infer::InferCtxt>::super_combine_tys::<rustc_infer[6cab06927fa701c3]::infer::equate::Equate>
  28:        0x117293289 - <rustc_infer[6cab06927fa701c3]::infer::equate::Equate as rustc_middle[7044477369b49eec]::ty::relate::TypeRelation>::tys
  29:        0x1186d5363 - <rustc_infer[6cab06927fa701c3]::infer::InferCtxt>::commit_if_ok::<rustc_infer[6cab06927fa701c3]::infer::InferOk<()>, rustc_middle[7044477369b49eec]::ty::error::TypeError, <rustc_infer[6cab06927fa701c3]::infer::at::Trace>::eq<rustc_middle[7044477369b49eec]::ty::Ty>::{closure#0}>
  30:        0x1186e3c70 - <rustc_infer[6cab06927fa701c3]::infer::InferCtxt>::can_eq::<rustc_middle[7044477369b49eec]::ty::Ty>
  31:        0x1186b184c - rustc_trait_selection[957489648007299]::traits::error_reporting::suggestions::hint_missing_borrow
  32:        0x1187bd1aa - <rustc_infer[6cab06927fa701c3]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[957489648007299]::traits::error_reporting::suggestions::TypeErrCtxtExt>::report_closure_arg_mismatch
  33:        0x1187d06f1 - <rustc_infer[6cab06927fa701c3]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[957489648007299]::traits::error_reporting::TypeErrCtxtExt>::report_selection_error
  34:        0x1187e0b2f - <rustc_infer[6cab06927fa701c3]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[957489648007299]::traits::error_reporting::InferCtxtPrivExt>::report_fulfillment_error
  35:        0x1187ca44b - <rustc_infer[6cab06927fa701c3]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[957489648007299]::traits::error_reporting::TypeErrCtxtExt>::report_fulfillment_errors
  36:        0x116fdb743 - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_argument_types
  37:        0x116fda7c8 - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_method_argument_types
  38:        0x117033fd8 - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_expr_kind
  39:        0x116fca6b5 - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  40:        0x11702c71b - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_expr_kind
  41:        0x116fca6b5 - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  42:        0x116fe373c - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_decl
  43:        0x116fe3c24 - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_stmt
  44:        0x116fe44df - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_block_with_expected
  45:        0x11702cfaf - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_expr_kind
  46:        0x116fca6b5 - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  47:        0x116fcc0e0 - <rustc_hir_typeck[6efceea4454c75c9]::fn_ctxt::FnCtxt>::check_return_expr
  48:        0x1170a7afc - rustc_hir_typeck[6efceea4454c75c9]::check::check_fn
  49:        0x1170e2670 - rustc_hir_typeck[6efceea4454c75c9]::typeck
  50:        0x118198a54 - <rustc_query_system[3d399bab0eaade7f]::dep_graph::graph::DepGraph<rustc_middle[7044477369b49eec]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[7044477369b49eec]::ty::context::TyCtxt, rustc_span[abe350e59e28df37]::def_id::LocalDefId, &rustc_middle[7044477369b49eec]::ty::typeck_results::TypeckResults>
  51:        0x117fa3286 - rustc_query_system[3d399bab0eaade7f]::query::plumbing::try_execute_query::<rustc_query_impl[469a71c953212c97]::queries::typeck, rustc_query_impl[469a71c953212c97]::plumbing::QueryCtxt>
  52:        0x1180c7b4e - <rustc_query_impl[469a71c953212c97]::Queries as rustc_middle[7044477369b49eec]::ty::query::QueryEngine>::typeck
  53:        0x11706dfab - rustc_data_structures[a3be787722aab228]::sync::par_for_each_in::<&[rustc_span[abe350e59e28df37]::def_id::LocalDefId], <rustc_middle[7044477369b49eec]::hir::map::Map>::par_body_owners<rustc_hir_typeck[6efceea4454c75c9]::typeck_item_bodies::{closure#0}>::{closure#0}>
  54:        0x1170e00a4 - rustc_hir_typeck[6efceea4454c75c9]::typeck_item_bodies
  55:        0x1181d4e59 - <rustc_query_system[3d399bab0eaade7f]::dep_graph::graph::DepGraph<rustc_middle[7044477369b49eec]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[7044477369b49eec]::ty::context::TyCtxt, (), ()>
  56:        0x117f089df - rustc_query_system[3d399bab0eaade7f]::query::plumbing::try_execute_query::<rustc_query_impl[469a71c953212c97]::queries::typeck_item_bodies, rustc_query_impl[469a71c953212c97]::plumbing::QueryCtxt>
  57:        0x1180c7a6f - <rustc_query_impl[469a71c953212c97]::Queries as rustc_middle[7044477369b49eec]::ty::query::QueryEngine>::typeck_item_bodies
  58:        0x116e580f7 - <rustc_session[d7fa5f284918d396]::session::Session>::time::<(), rustc_hir_analysis[27a8f7e0703a01ee]::check_crate::{closure#7}>
  59:        0x116df4cb3 - rustc_hir_analysis[27a8f7e0703a01ee]::check_crate
  60:        0x117350079 - rustc_interface[d2d1d1fbee7644ce]::passes::analysis
  61:        0x1181cae99 - <rustc_query_system[3d399bab0eaade7f]::dep_graph::graph::DepGraph<rustc_middle[7044477369b49eec]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[7044477369b49eec]::ty::context::TyCtxt, (), core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>>
  62:        0x117fa613b - rustc_query_system[3d399bab0eaade7f]::query::plumbing::try_execute_query::<rustc_query_impl[469a71c953212c97]::queries::analysis, rustc_query_impl[469a71c953212c97]::plumbing::QueryCtxt>
  63:        0x1180c35a8 - <rustc_query_impl[469a71c953212c97]::Queries as rustc_middle[7044477369b49eec]::ty::query::QueryEngine>::analysis
  64:        0x116b3396a - <rustc_interface[d2d1d1fbee7644ce]::passes::QueryContext>::enter::<rustc_driver[38920cb2344860e2]::run_compiler::{closure#1}::{closure#2}::{closure#2}, core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>>
  65:        0x116b14f5d - <rustc_interface[d2d1d1fbee7644ce]::queries::QueryResult<rustc_interface[d2d1d1fbee7644ce]::passes::QueryContext>>::enter::<core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>, rustc_driver[38920cb2344860e2]::run_compiler::{closure#1}::{closure#2}::{closure#2}>
  66:        0x116b72da3 - <rustc_interface[d2d1d1fbee7644ce]::interface::Compiler>::enter::<rustc_driver[38920cb2344860e2]::run_compiler::{closure#1}::{closure#2}, core[fd5e363bb4760a86]::result::Result<core[fd5e363bb4760a86]::option::Option<rustc_interface[d2d1d1fbee7644ce]::queries::Linker>, rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>>
  67:        0x116b8fa86 - rustc_span[abe350e59e28df37]::with_source_map::<core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>, rustc_interface[d2d1d1fbee7644ce]::interface::run_compiler<core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>, rustc_driver[38920cb2344860e2]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  68:        0x116b3a3e5 - std[963c226d0575182f]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[d2d1d1fbee7644ce]::util::run_in_thread_pool_with_globals<rustc_interface[d2d1d1fbee7644ce]::interface::run_compiler<core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>, rustc_driver[38920cb2344860e2]::run_compiler::{closure#1}>::{closure#0}, core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>>
  69:        0x116b19385 - <<std[963c226d0575182f]::thread::Builder>::spawn_unchecked_<rustc_interface[d2d1d1fbee7644ce]::util::run_in_thread_pool_with_globals<rustc_interface[d2d1d1fbee7644ce]::interface::run_compiler<core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>, rustc_driver[38920cb2344860e2]::run_compiler::{closure#1}>::{closure#0}, core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[fd5e363bb4760a86]::result::Result<(), rustc_errors[112d18aeec5719f8]::ErrorGuaranteed>>::{closure#1} as core[fd5e363bb4760a86]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  70:        0x10cf8e5a7 - std::sys::unix::thread::Thread::new::thread_start::had0599fead9033cd
  71:     0x7ff80954b1d3 - __pthread_start

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.68.2 (9eb3afe9e 2023-03-27) running on x86_64-apple-darwin

note: compiler flags: --crate-type bin -C embed-bitcode=no -C split-debuginfo=unpacked -C debuginfo=2 -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `main`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `count_words`
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
 Compiling count_words v0.1.0 (/Users/harrylaou/myworkspace/rust/linkedin-rust-essential-training/count_words)
error: internal compiler error: compiler/rustc_infer/src/infer/region_constraints/mod.rs:568:17: cannot relate bound region: ReLateBound(DebruijnIndex(0), BoundRegion { var: 1, kind: BrNamed(DefId(0:9 ~ count_words[0ce7]::main::compare::'_#1), '_) }) <= '_#20r

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/compiler/rustc_errors/src/lib.rs:987:33
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: std::panic::panic_any::<rustc_errors::ExplicitBug>
   2: <rustc_errors::HandlerInner>::span_bug::<rustc_span::span_encoding::Span, &alloc::string::String>
   3: <rustc_errors::Handler>::span_bug::<rustc_span::span_encoding::Span, &alloc::string::String>
   4: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>::{closure#0}
   5: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   6: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   7: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>
   8: rustc_middle::util::bug::span_bug_fmt::<rustc_span::span_encoding::Span>
   9: <rustc_infer::infer::region_constraints::RegionConstraintCollector>::make_subregion
  10: <rustc_infer::infer::region_constraints::RegionConstraintCollector>::make_eqregion
  11: <rustc_infer::infer::equate::Equate as rustc_middle::ty::relate::TypeRelation>::regions
  12: rustc_middle::ty::relate::super_relate_tys::<rustc_infer::infer::equate::Equate>
  13: <rustc_infer::infer::InferCtxt>::super_combine_tys::<rustc_infer::infer::equate::Equate>
  14: <rustc_infer::infer::equate::Equate as rustc_middle::ty::relate::TypeRelation>::tys
  15: <core::result::Result<rustc_middle::ty::Ty, rustc_middle::ty::error::TypeError> as rustc_type_ir::InternIteratorElement<rustc_middle::ty::Ty, rustc_middle::ty::Ty>>::intern_with::<core::iter::adapters::map::Map<core::iter::adapters::zip::Zip<core::iter::adapters::copied::Copied<core::slice::iter::Iter<rustc_middle::ty::Ty>>, core::iter::adapters::copied::Copied<core::slice::iter::Iter<rustc_middle::ty::Ty>>>, rustc_middle::ty::relate::super_relate_tys<rustc_infer::infer::equate::Equate>::{closure#2}>, <rustc_middle::ty::context::TyCtxt>::mk_tup<core::iter::adapters::map::Map<core::iter::adapters::zip::Zip<core::iter::adapters::copied::Copied<core::slice::iter::Iter<rustc_middle::ty::Ty>>, core::iter::adapters::copied::Copied<core::slice::iter::Iter<rustc_middle::ty::Ty>>>, rustc_middle::ty::relate::super_relate_tys<rustc_infer::infer::equate::Equate>::{closure#2}>>::{closure#0}>
  16: rustc_middle::ty::relate::super_relate_tys::<rustc_infer::infer::equate::Equate>
  17: <rustc_infer::infer::InferCtxt>::super_combine_tys::<rustc_infer::infer::equate::Equate>
  18: <rustc_infer::infer::equate::Equate as rustc_middle::ty::relate::TypeRelation>::tys
  19: <rustc_infer::infer::InferCtxt>::commit_if_ok::<rustc_infer::infer::InferOk<()>, rustc_middle::ty::error::TypeError, <rustc_infer::infer::at::Trace>::eq<rustc_middle::ty::Ty>::{closure#0}>
  20: <rustc_infer::infer::InferCtxt>::can_eq::<rustc_middle::ty::Ty>
  21: rustc_trait_selection::traits::error_reporting::suggestions::hint_missing_borrow
  22: <rustc_infer::infer::error_reporting::TypeErrCtxt as rustc_trait_selection::traits::error_reporting::suggestions::TypeErrCtxtExt>::report_closure_arg_mismatch
  23: <rustc_infer::infer::error_reporting::TypeErrCtxt as rustc_trait_selection::traits::error_reporting::TypeErrCtxtExt>::report_selection_error
  24: <rustc_infer::infer::error_reporting::TypeErrCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtPrivExt>::report_fulfillment_error
  25: <rustc_infer::infer::error_reporting::TypeErrCtxt as rustc_trait_selection::traits::error_reporting::TypeErrCtxtExt>::report_fulfillment_errors
  26: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_argument_types
  27: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_method_argument_types
  28: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  29: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  30: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  31: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  32: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_decl
  33: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_stmt
  34: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_block_with_expected
  35: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  36: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  37: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_return_expr
  38: rustc_hir_typeck::check::check_fn
  39: rustc_hir_typeck::typeck
  40: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, rustc_span::def_id::LocalDefId, &rustc_middle::ty::typeck_results::TypeckResults>
  41: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::queries::typeck, rustc_query_impl::plumbing::QueryCtxt>
  42: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck
  43: rustc_data_structures::sync::par_for_each_in::<&[rustc_span::def_id::LocalDefId], <rustc_middle::hir::map::Map>::par_body_owners<rustc_hir_typeck::typeck_item_bodies::{closure#0}>::{closure#0}>
  44: rustc_hir_typeck::typeck_item_bodies
  45: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), ()>
  46: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::queries::typeck_item_bodies, rustc_query_impl::plumbing::QueryCtxt>
  47: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck_item_bodies
  48: <rustc_session::session::Session>::time::<(), rustc_hir_analysis::check_crate::{closure#7}>
  49: rustc_hir_analysis::check_crate
  50: rustc_interface::passes::analysis
  51: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), core::result::Result<(), rustc_errors::ErrorGuaranteed>>
  52: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
  53: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis
  54: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#2}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
  55: <rustc_interface::queries::QueryResult<rustc_interface::passes::QueryContext>>::enter::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#2}>
  56: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorGuaranteed>>
  57: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.68.2 (9eb3afe9e 2023-03-27) running on x86_64-apple-darwin

note: compiler flags: --crate-type bin -C embed-bitcode=no -C split-debuginfo=unpacked -C debuginfo=2 -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `main`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `count_words`
```

</p>
</details>

**Comment by lukas-code** [comment]
This is fixed in the latest stable version (1.69.0), but you're still running 1.68.2.

Old stable versions generally don't get patches, so you should probably update to the latest version.

If you installed rust with rustup, you can get the latest release by running `rustup update`.

**Comment by Nilstrieb** [comment]
Thank you for the report anyways!
